### PR TITLE
enable PublicApiAnalyzer for shipping assemblies

### LINKED ISSUE #3105: [public API] Configure public API code analyzer to prevent non-intentional API changes

### PR DESCRIPTION
**Pull request by vlada-shubina**
### Problem
fixes #3105 

### Solution
enables PublicApiAnalyzer for shipping assemblies
fixes API issues detected by analyzer

PS: shipping files should be filled before preview branding.

### Checks:
- [x] Added unit tests - NA
- [x] Added `#nullable enable` to all the modified files [?](https://github.com/dotnet/templating/wiki/Contributing#coding-style)

### CODE REVIEW

File: eng/Analyzers.props [hunk 1]
```diff
@@ -10,4 +10,10 @@
     <PackageReference Include="Microsoft.CodeAnalysis.CSharp.CodeStyle" Version="3.8.0" PrivateAssets="all" />
     <PackageReference Include="StyleCop.Analyzers" Version="1.2.0-beta.333" PrivateAssets="all" />
   </ItemGroup>
+  <ItemGroup Condition="'$(EnableAnalyzers)' == 'true' And '$(EnablePublicApiAnalyzer)' == 'true'">
+    <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="3.3.2">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+  </ItemGroup>
```
Comment (DavidKarlas) [inline eng/Analyzers.props:18]: I believe all this could be single line like this:
`<PackageReference Condition="'$(EnablePublicApiAnalyzer)' == 'true'" Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="3.3.2" PrivateAssets="all" IncludeAssets="runtime; build; native; contentfiles; analyzers; buildtransitive" >`
in above section.

File: src/Microsoft.TemplateEngine.Core/Microsoft.TemplateEngine.Core.csproj [hunk 2]
```diff
@@ -1,14 +1,21 @@
...
+    <Description>Common operations for instantiating templates using forward-only input stream operations</Description>
+    <IsPackable>true</IsPackable>
+    <EnablePublicApiAnalyzer>true</EnablePublicApiAnalyzer>
+  </PropertyGroup>
+  <ItemGroup Condition="'$(EnableAnalyzers)' == 'true'">
+    <PackageReference Include="Microsoft.CodeAnalysis.PublicApiAnalyzers" Version="3.3.2">
+      <PrivateAssets>all</PrivateAssets>
+      <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
+    </PackageReference>
+  </ItemGroup>
```
Comment (DavidKarlas) [inline src/Microsoft.TemplateEngine.Core/Microsoft.TemplateEngine.Core.csproj:14]: Why is this needed? Isn't Analyzers.props supposed to insert this?

**Review by DavidKarlas — APPROVED** [review]

File: eng/CodeAnalysis.ruleset [hunk 3]
```diff
@@ -486,4 +486,7 @@
     <Rule Id="IDE1007" Action="Hidden" />       <!-- UnboundIdentifier -->
     <Rule Id="IDE1008" Action="Hidden" />       <!-- UnboundConstructor -->
   </Rules>
+  <Rules AnalyzerId="Microsoft.CodeAnalysis.PublicApiAnalyzers" RuleNamespace="Microsoft.CodeAnalysis.PublicApiAnalyzers" >
+    <Rule Id="RS0041" Action="Info" TESpecific="true" /> <!--Public members should not use oblivious types -->
```
[the remark this answers is not in the document]
Comment (vlada-shubina) [reply eng/CodeAnalysis.ruleset:490]: the issue is that we have nullable enable only in some files. This rule doesn't allow mixture when nullables are enabled.

**Comment by vlada-shubina** [comment]
@bozturkMSFT @DavidKarlas please re-review https://github.com/dotnet/templating/pull/3145/commits/6175b4afc6fc073758bb08067f10537d6eb6999c

**Review by bozturkMSFT — APPROVED** [review]

File: eng/CodeAnalysis.ruleset (hunk 3 above)
[the remark this answers is not in the document]
Comment (bozturkMSFT) [reply eng/CodeAnalysis.ruleset]: I find that very strange. Having `nullable enable` only in some files is the only usage of that directive, otherwise we would just enable it in csproj 😄
